src/ControlSystem/Trigger.hpp

```diff
@@ -148,15 +161,53 @@ class Trigger : public DenseTrigger {
       const std::unordered_map<
           std::string,
           std::unique_ptr<domain::FunctionsOfTime::FunctionOfTime>>&
-          measurement_timescales) {
-    return time +
-           tmpl::as_pack<ControlSystems>(
-               [&measurement_timescales, &time](auto... control_systems) {
-                 return std::min({min(
-                     measurement_timescales
-                         .at(tmpl::type_from<decltype(control_systems)>::name())
-                         ->func(time)[0])...});
-               });
+          measurement_timescales,
+      const std::unordered_map<
+          std::string,
+          std::unique_ptr<domain::FunctionsOfTime::FunctionOfTime>>&
+          functions_of_time) {
+    const double min_measure_time = tmpl::as_pack<ControlSystems>(
+        [&measurement_timescales, &time](auto... control_systems) {
+          return std::min(
+              {min(measurement_timescales
+                       .at(tmpl::type_from<decltype(control_systems)>::name())
+                       ->func(time)[0])...});
+        });
+    if (min_measure_time == std::numeric_limits<double>::infinity()) {
+      return min_measure_time;
+    }
+    const double calculated_next_measurement = time + min_measure_time;
+
+    double min_fot_expr_time = std::numeric_limits<double>::infinity();
+    for (const auto& [name, fot] : functions_of_time) {
+      (void)name;
+      min_fot_expr_time = std::min(min_fot_expr_time, fot->time_bounds()[1]);
+    }
+
+    // If the expiration time is infinity, don't bother checking if it's equal
+    // to the min measure time. We already returned above if it is infinity
+    if (min_fot_expr_time == std::numeric_limits<double>::infinity()) {
+      return calculated_next_measurement;
+    }
+
+    double next_measurement_time;
+    // Just take the average of the two things we are comparing. They'll always
+    // be super close
+    const double scale =
+        0.5 * (calculated_next_measurement + min_fot_expr_time);
```

Comment:
Pass zero as the scale.  The scale is to correct for error caused by intermediate results canceling in the calculation.